src/main/java/com/epam/aidial/core/security/ApiKeyStore.java

```diff
@@ -2,57 +2,117 @@
 
 import com.epam.aidial.core.config.ApiKeyData;
 import com.epam.aidial.core.config.Key;
+import com.epam.aidial.core.data.ResourceType;
+import com.epam.aidial.core.service.LockService;
+import com.epam.aidial.core.service.ResourceService;
+import com.epam.aidial.core.storage.ResourceDescription;
+import com.epam.aidial.core.util.ProxyUtil;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
 import java.util.HashMap;
-import java.util.Iterator;
 import java.util.Map;
+import javax.annotation.concurrent.GuardedBy;
 
 import static com.epam.aidial.core.security.ApiKeyGenerator.generateKey;
+import static com.epam.aidial.core.storage.BlobStorageUtil.PATH_SEPARATOR;
 
 @Slf4j
+@AllArgsConstructor
 public class ApiKeyStore {
 
+    public static final String API_KEY_DATA_BUCKET = "api_key_data";
+    public static final String API_KEY_DATA_LOCATION = API_KEY_DATA_BUCKET + PATH_SEPARATOR;
+
+    private final ResourceService resourceService;
+
+    private final LockService lockService;
+
+    private final Vertx vertx;
+
+    /**
+     * API keys are captured from secure storage.
+     */
+    @GuardedBy("this")
     private final Map<String, ApiKeyData> keys = new HashMap<>();
 
     public synchronized void assignApiKey(ApiKeyData data) {
-        String apiKey = generateApiKey();
-        keys.put(apiKey, data);
-        data.setPerRequestKey(apiKey);
+        lockService.underBucketLock(API_KEY_DATA_LOCATION, () -> {
+            ResourceDescription resource = generateApiKey();
+            String apiKey = resource.getName();
+            data.setPerRequestKey(apiKey);
+            String json = ProxyUtil.convertToString(data);
+            if (resourceService.putResource(resource, json, false, false) == null) {
+                throw new IllegalStateException(String.format("API key %s already exists in the storage", apiKey));
+            }
+            return apiKey;
+        });
     }
 
-    public synchronized ApiKeyData getApiKeyData(String key) {
-        return keys.get(key);
+    public synchronized Future<ApiKeyData> getApiKeyData(String key) {
+        ApiKeyData apiKeyData = keys.get(key);
+        if (apiKeyData != null) {
+            return Future.succeededFuture(apiKeyData);
+        }
+        ResourceDescription resource = toResource(key);
+        return vertx.executeBlocking(() -> ProxyUtil.convertToObject(resourceService.getResource(resource), ApiKeyData.class));
     }
 
-    public synchronized void invalidateApiKey(ApiKeyData apiKeyData) {
-        if (apiKeyData.getPerRequestKey() != null) {
-            keys.remove(apiKeyData.getPerRequestKey());
+    public Future<Boolean> invalidateApiKey(ApiKeyData apiKeyData) {
```

Comment:
does invalidation require bucket lock as well?

Reply:
I think it doesn't unless you give an example where that might be required.

Reply:
You are right, my initial intention was to block access to this key during finalization. but now I realized that it can't be solved by just adding bucket lock for invalidation, because getApiKey does not hold any lock. let's leave it as is